packages/elements/src/react/common/form/otp.tsx

```diff
@@ -45,120 +83,314 @@ export const OTPInput = React.forwardRef<HTMLInputElement, OTPInputProps>(functi
-    const [selectionRange, setSelectionRange] = React.useState<SelectionRange>(props.autoFocus ? ZERO : OUTSIDE);
+    const value = props.value as string;
+    const previousValue = usePrevious(value);
 
+    const innerRef = React.useRef<HTMLInputElement>(null);
     // This ensures we can access innerRef internally while still exposing it via the ref prop
     React.useImperativeHandle(ref, () => innerRef.current as HTMLInputElement, []);
 
-    // A layout effect is used here to avoid any perceived visual lag when changing the selection
-    React.useLayoutEffect(() => {
-      if (document.activeElement !== innerRef.current) {
-        return;
-      }
-      setSelectionRange(cur => selectionRangeUpdater(cur, innerRef));
-    }, [props.value]);
+    const [selectionRange, setSelectionRange] = React.useState<SelectionRange>(props.autoFocus ? ZERO : OUTSIDE);
+    const isFocused = useFocus(innerRef);
+
+    // Attach this component's UI state "selectionRange" to the native .select() to update it on focus changes
+    React.useEffect(() => {
+      const element = innerRef.current;
+
+      if (!element) return;
+
+      const originalSelect = element.select.bind(element);
+      element.select = () => {
+        originalSelect();
+        setSelectionRange([0, element.value.length]);
+      };
+    }, []);
+
+    // Continuously update the selection range when the input is focused
+    React.useEffect(() => {
+      if (!isFocused) return;
+
+      const interval = setInterval(() => {
+        if (innerRef.current && document.activeElement === innerRef.current) {
+          setSelectionRange([innerRef.current.selectionStart, innerRef.current.selectionEnd]);
+        }
+      }, 50);
+
+      return () => {
+        clearInterval(interval);
+      };
+    }, [isFocused]);
 
     // Fire the requestSubmit callback when the input has the required length and autoSubmit is enabled
     React.useEffect(() => {
-      if (String(props.value).length === length && autoSubmit) {
+      if (previousValue === undefined) return;
+
+      if (value !== previousValue && previousValue.length < length && value.length === length && autoSubmit) {
         innerRef.current?.form?.requestSubmit();
       }
-    }, [props.value, length, autoSubmit]);
+    }, [previousValue, value, length, autoSubmit]);
+
+    const _select = React.useCallback(() => {
+      if (!innerRef.current) return;
+
+      const [start, end] = transformSelectionRange({ innerRef, length, value });
+      setSelectionRange([start, end]);
+    }, [innerRef, value, length]);
+
+    const _keyDown = React.useCallback(
+      (e: React.KeyboardEvent<HTMLInputElement>) => {
+        transformKeydownEvent({ e, innerRef, value });
+      },
+      [innerRef, value],
+    );
 
     return (
       <div
-        style={
-          {
-            position: 'relative',
-          } as React.CSSProperties
-        }
+        data-otp-input-segmented-wrapper
+        style={wrapperStyle}
```

Comment:
Moved the styles from inline to dedicated objects (throughout). Also added data attributes where applicable to allow for CSS styling